system/sqlstore.go

```diff
@@ -0,0 +1,194 @@
+			}
+		}
+	}()
+	return s
+}
+
+// Get returns a cached session.
+func (s *SQLStore) Get(r *http.Request, name string) (*sessions.Session, error) {
+	return sessions.GetRegistry(r).Get(s, name)
+}
+
+// New creates a new session for the given request r. If the request
+// contains a valid session ID for an existing, non-expired session,
+// then that session will be loaded from the database.
+func (s *SQLStore) New(r *http.Request, name string) (*sessions.Session, error) {
+	session := sessions.NewSession(s, name)
+	opts := *s.Options
+	session.Options = &opts
+	c, err := r.Cookie(name)
+	if err != nil {
+		if err == http.ErrNoCookie {
+			return session, nil
+		}
+		return session, err
+	}
+	err = securecookie.DecodeMulti(name, c.Value, &session.ID, s.codecs...)
+	if err != nil {
+		// these are not the sessions you are looking for
+		return session, nil
+	}
+	err = s.load(session)
+	if err != nil {
+		return session, err
+	}
+	return session, err
+}
+
+// Save stores the session in the database. If session.Options.MaxAge
+// is < 0, the session is deleted from the database.
+func (s *SQLStore) Save(r *http.Request, w http.ResponseWriter, session *sessions.Session) error {
+	if session.Options.MaxAge < 0 {
+		return s.destroy(session)
+	}
+	if len(session.ID) == 0 {
+		session.ID = base32.StdEncoding.EncodeToString(securecookie.GenerateRandomKey(32))
+	}
+	if err := s.save(session); err != nil {
+		return err
+	}
+	// data is not stored in the cookie, only the session id
+	encoded, err := securecookie.EncodeMulti(session.Name(), &session.ID, s.codecs...)
+	if err != nil {
+		return err
+	}
+	http.SetCookie(w, sessions.NewCookie(session.Name(), encoded, session.Options))
+	return nil
+}
+
+// load loads the session identified by its ID from the database if it
+// exists. If the session has expired, it is destroyed.
+func (s *SQLStore) load(session *sessions.Session) error {
+	var dbSession models.Session
+	if err := s.dbMap.SelectOne(&dbSession, "SELECT * FROM Session WHERE Token = ?", session.ID); err != nil {
+		// if no rows are found nothing is done
+		if err == sql.ErrNoRows {
+			return nil
+		}
+		return fmt.Errorf("Could not select session to destroy: %v", err)
+	}
+	if dbSession.Expires < time.Now().Unix() {
+		return s.destroy(session)
+	}
+	// write db Data to session.Values
+	return gob.NewDecoder(bytes.NewBuffer(dbSession.Data)).Decode(&session.Values)
+}
+
+// save checks whether the session is new and inserts if new. Updates if
+// not.
+func (s *SQLStore) save(session *sessions.Session) error {
+	var dbSession models.Session
+	var buf bytes.Buffer
+	var isNew bool
+	if err := s.dbMap.SelectOne(&dbSession, "SELECT * FROM Session WHERE Token = ?", session.ID); err != nil {
+		if err != sql.ErrNoRows {
+			return fmt.Errorf("Could not select session: %v", err)
+		}
+		// no rows found so new
+		isNew = true
+	}
+	if userID, ok := session.Values["UserId"].(int64); ok {
+		dbSession.UserId = userID
+	} else {
+		dbSession.UserId = -1
```

Comment:
I am defining no user as -1.  I wasn't sure if there could be a user 0, and I also don't think gorp can handle a nil value.  Since there can never be a user in the negatives, it just means a session where the user has not signed in yet, or they aren't registered. Unless I am wrong, I will put a comment here because it is not intuitive.